fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/RecordQueryPlanner.java

```diff
@@ -406,7 +407,11 @@ private ScoredPlan planFilter(@Nonnull PlanContext planContext, @Nonnull QueryCo
             boolean canSort = inExtractor.setSort(planContext.query.getSort(), planContext.query.isSortReverse());
             if (!canSort && getConfiguration().shouldAttemptFailedInJoinAsOr()) {
                 // Can't implement as an in join because of the sort order. Try as an OR instead.
-                withInAsOr = planFilter(planContext, normalizeAndOrForInAsOr(inExtractor.asOr()));
+                // Only try if extracting the OR actually changes the filter to avoid recursing infinitely.
+                final QueryComponent asOr = normalizeAndOrForInAsOr(inExtractor.asOr());
+                if (!filter.equals(asOr)) {
```

Comment:
I think this will do what you have in mind, although it's possibly better to make this dependent on the state of the `inExtractor`. We shouldn't attempt it if the `inClauses` in there were non-empty before doing `setSort()`. If that's the case `asOr()` will actually do work and create a new modified filter replacing the IN with and OR construction.

Reply:
I think it's little more subtle than the "in" clauses being non-empty, as we actually already check the emptiness in of the in clauses in `setSort` (and return `true`, so won't go down the path that was infinitely recursing). The somewhat more fundamental issue is that the `in` clauses can't be transformed if they are all `InParameterClause`s, so we could check that (instead of checking to see if things get modified). I have a slight preference for this method, as it seems slightly more resilient to future modifications that introduce other mechanisms for `asOr` improving or changing, but it's not something I'm super committed to either way.

Sorry, it took a while to respond; got distracted.